common/djangoapps/track/transformers.py

```diff
@@ -0,0 +1,484 @@
+    """
+    mapping = DottedPathMapping()
+
+    @classmethod
+    def register(cls, transformer):
+        """
+        Decorator to register an EventTransformer.  It must have a `match_key`
+        class attribute defined.
+        """
+        cls.mapping[transformer.match_key] = transformer
+        return transformer
+
+    @classmethod
+    def create_transformer(cls, event):
+        """
+        Create an EventTransformer of the given event.
+
+        If no transformer is registered to handle the event, this raises a
+        KeyError.
+        """
+        name = event.get(u'name')
+        return cls.mapping[name](event)
+
+
+class EventTransformer(dict):
+    """
+    Creates a transformer to modify analytics events based on event type.
+
+    To use the transformer, instantiate it using the
+    `EventTransformer.create_transformer()` classmethod with the event
+    dictionary as the sole argument, and then call `transformer.transform()` on
+    the created object to modify the event to the format required for output.
+
+    Custom transformers will want to define some or all of the following values
+
+    Attributes:
+
+        match_key:
+            This is the name of the event you want to transform.  If the name
+            ends with a `'.'`, it will be treated as a *prefix transformer*.
+            All other names denote *exact transformers*.
+
+            A *prefix transformer* will handle any event whose name begins with
+            the name of the prefix transformer.  Only the most specific match
+            will be used, so if a transformer exists with a name of
+            `'edx.ui.lms.'` and another transformer has the name
+            `'edx.ui.lms.sequence.'` then an event called
+            `'edx.ui.lms.sequence.tab_selected'` will be handled by the
+            `'edx.ui.lms.sequence.'` transformer.
+
+            An *exact transformer* will only handle events whose name matches
+            name of the transformer exactly.
+
+            Exact transformers always take precedence over prefix transformers.
+
+            Transformers without a name will not be added to the registry, and
+            cannot
+            be accessed via the `EventTransformer.create_transformer()`
+            classmethod.
+
+        is_legacy_event:
+            If an event is a legacy event, it needs to set event_type to the
+            legacy name for the event, and may need to set certain event fields
+            to maintain backward compatiblity.  If an event needs to provide
+            legacy support in some contexts, `is_legacy_event` can be defined
+            as a property to add dynamic behavior.
+
+            Default: False
+
+        legacy_event_type:
+            If the event is or can be a legacy event, it should define
+            the legacy value for the event_type field here.
+
+    Processing methods.  Override these to provide the behavior needed for your
```

Comment:
I went back and forth on these.  They are used internally, but they are also the primary API for the Transformer.  So similar to how Django CBVs have a `get()` and `post()` method that are called internally (by `View.dispatch()`), but are not underscore-prefixed, because they are the main way developers interact with the class.  Subclasses will generally override these one or both of these two methods, so I sided with leaving them un-prefixed for mostly aesthetic reasons, but I'm not wedded to it.  Happy to prefix them.